[WPE] Variable length array build warning

philn opened the pull request:
#### 72ef47e183a417a7e337b9b5b7d5ae448079f744
<pre>
[WPE] Variable length array build warning
<a href="https://bugs.webkit.org/show_bug.cgi?id=272830">https://bugs.webkit.org/show_bug.cgi?id=272830</a>

Reviewed by Michael Catanzaro.

Variable-length arrays (VLAs) are supported in C (since the C99 standard). But not in the C++
standard. This is an extension of some compilers.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDRMCursor.cpp:
(WPE::DRM::Cursor::updateBuffer): Store the cursor pixels data in a Vector.
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp: Store the viewConfigurations in a Vector.

Canonical link: <a href="https://commits.webkit.org/277675@main">https://commits.webkit.org/277675@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/268799560eabfabb27f3fbeb3709a77b4783d2b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27380 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50860 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44237 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24891 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39361 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48753 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25088 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41645 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20494 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22566 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42821 "Found 1 new test failure: fast/css/viewport-height-outline.html (failure)") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6228 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44538 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43265 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52763 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23218 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19581 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46696 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/capture-with-visibility-hidden-child.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45591 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10649 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24206 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDRMCursor.cpp:
```diff
@@ -69,11 +69,11 @@ void Cursor::updateBuffer(const uint8_t* pixels, uint32_t width, uint32_t height
-        memcpy(deviceBuffer + i * m_deviceWidth, pixels + i * stride, stride);
-    gbm_bo_write(m_buffer->bufferObject(), deviceBuffer, sizeof(deviceBuffer));
+        memcpy(&deviceBuffer[i * m_deviceWidth], pixels + i * stride, stride);
+    gbm_bo_write(m_buffer->bufferObject(), deviceBuffer.data(), deviceBuffer.sizeInBytes());
```
Hm, does this really work? The capacity of the vector is not the same as its size. No elements are added to the vector, so even if setting the capacity is sufficient to avoid memory corruption, I assume its size is still zero at the end here? I think you want to instead set the initial size. `Vector<uint32_t> deviceBuffer(m_deviceWidth * m_deviceHeight);` should get you a vector of the correct size with uninitialized values.

philn replied:
Ah yes I think you're right.

mcatanzaro reviewed (approved):
I'm going to click Approve now, but you might be able to make the `memcpy` safer by using `deviceBuffer.subspan()` and `memsetSpan` rather than raw `memcpy`.

philn committed:
[WPE] Variable length array build warning
https://bugs.webkit.org/show_bug.cgi?id=272830

Reviewed by Michael Catanzaro.

Variable-length arrays (VLAs) are supported in C (since the C99 standard). But not in the C++
standard. This is an extension of some compilers.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDRMCursor.cpp:
(WPE::DRM::Cursor::updateBuffer): Store the cursor pixels data in a Vector.
* Source/WebCore/platform/xr/openxr/PlatformXROpenXR.cpp: Store the viewConfigurations in a Vector.

Canonical link: https://commits.webkit.org/277675@main
